### PR TITLE
Feature: Number Prop Pointer Lock drag

### PR DESCRIPTION
**Pull request by vezwork**
- Used the [pointer lock API](https://mdn.github.io/dom-examples/pointer-lock/) to lock the cursor when dragging on the number prop editor. This eliminates the porblem where a user is adjusting the number prop by dragging on the editor and their mouse hits the side of the screen.

Note: I have disabled this behaviour on Safari. @AndrewPrifer showed us that safari reflows the entire page with a banner when pointer locking which is not good UX.

https://user-images.githubusercontent.com/11082236/167917628-651eeeaf-c225-4920-80a8-2c156cd2631c.mp4

## Testing notes:
- For each:
  - [x] Chrome
  - [x] Firefox
  - [x] Safari
- tested the following:
  - [x] I can drag a number input with a range
  - [x] I can drag a number input without a range 
  - [x] I can click a number input and edit it as text (no regression)

### CODE REVIEW

File: theatre/studio/src/uiComponents/DraggableArea.tsx [hunk 1]
```diff
@@ -1,176 +1,13 @@
...
-  componentWillUnmount() {
-    if (this.mode !== 'notDragging') {
-      this.relinquishCursorLock()
-      this.relinquishCursorLock = voidFn
-      this.removeDragListeners()
-      this.props.onDragEnd && this.props.onDragEnd(this.mode === 'dragging')
-      this.mode = 'notDragging'
-    }
-  }
+  return <div ref={ref}>{props.children}</div>
```
[the remark this answers is not in the document]
Comment (vezwork) [reply theatre/studio/src/uiComponents/DraggableArea.tsx]: DraggableArea was deleted

File: theatre/studio/src/uiComponents/form/BasicNumberInput.tsx [hunk 2]
```diff
@@ -223,53 +221,51 @@ const BasicNumberInput: React.FC<{
...
-          magnitude: 1,
-        })
+      bodyCursorBeforeDrag.current = document.body.style.cursor
 
-      if (propsA.range) {
-        newValue = clamp(newValue, propsA.range[0], propsA.range[1])
+      return {
+        // note: we use mx because dx is always 0 with `shouldPointerLock={true}`
+        onDrag(_dx: number, _dy: number, _e: MouseEvent, mx: number) {
+          totalMovement += mx
```
[the remark this answers is not in the document]
Comment (vezwork) [reply theatre/studio/src/uiComponents/form/BasicNumberInput.tsx]: mx allows us to calculate the changing value and constrain it as the dragging happens

File: theatre/studio/src/panels/SequenceEditorPanel/RightOverlay/HorizontalScrollbar.tsx [hunk 3]
```diff
@@ -174,117 +175,143 @@ const HorizontalScrollbar: React.FC<{
...
+    lockCSSCursorTo: 'w-resize',
+  })
+
+  const [refRangeEndDrag, nodeRangeEndDrag] =
+    useRefAndState<HTMLDivElement | null>(null)
+  useDrag(nodeRangeEndDrag, {
+    debugName: 'HorizontalScrollbar DraggableArea onRangeEndDrag',
+    onDragStart: handles.onRangeEndDragStart,
+    lockCSSCursorTo: 'e-resize',
+  })
```
Comment (colelawrence) [inline theatre/studio/src/panels/SequenceEditorPanel/RightOverlay/HorizontalScrollbar.tsx:284]: Praise: Nice after being able to completely remove `DraggableArea`

File: theatre/studio/src/uiComponents/form/BasicNumberInput.tsx [hunk 4]
```diff
@@ -329,18 +320,18 @@ const BasicNumberInput: React.FC<{
     />
   ) : null
 
+  const [refDrag, nodeDrag] = useRefAndState<HTMLDivElement | null>(null)
+  useDrag(nodeDrag, {
+    debugName: 'basicNumberInput',
```
Comment (colelawrence) [inline theatre/studio/src/uiComponents/form/BasicNumberInput.tsx:325]: Praise: Nice to be able to now include a `debugName`

Comment (colelawrence) [inline theatre/studio/src/uiComponents/form/BasicNumberInput.tsx:325]: Nit: Consider `"form/BasicNumberInput"`

**Review by colelawrence — APPROVED** [review]

**Comment by vezwork** [comment]
@AriaMinaei fixed and tested on [latest deployment](https://theatre-playground-7h6hbtp6r-theatrejs.vercel.app/).

Testing notes:
For each
- [ ] Chrome
- [ ] Firefox
- [ ] Safari
I can
- [ ] Click on a number prop to begin editing it as text
- [ ] drag a number prop, and on firefox and chrome my pointer will be locked
- [ ] drag "past the edge of my screen"
- [ ] drag and the value of the number prop is constrained to within its max and min bounds. I **cannot** overshoot the bounds in such a way that when I move my mouse while dragging nothing happens.
- [ ] drag the horizontal scroll bar on the dope sheet

File: theatre/studio/src/panels/SequenceEditorPanel/RightOverlay/HorizontalScrollbar.tsx [hunk 5]
```diff
@@ -174,117 +175,143 @@ const HorizontalScrollbar: React.FC<{
...
+    lockCSSCursorTo: 'w-resize',
+  })
+
+  const [refRangeEndDrag, nodeRangeEndDrag] =
+    useRefAndState<HTMLDivElement | null>(null)
+  useDrag(nodeRangeEndDrag, {
+    debugName: 'HorizontalScrollbar/onRangeEndDrag',
+    onDragStart: handles.onRangeEndDragStart,
+    lockCSSCursorTo: 'e-resize',
+  })
```
Comment (colelawrence) [inline theatre/studio/src/panels/SequenceEditorPanel/RightOverlay/HorizontalScrollbar.tsx:284]: @AndrewPrifer, is this where we can just `useState`?

Comment (vezwork) [reply theatre/studio/src/panels/SequenceEditorPanel/RightOverlay/HorizontalScrollbar.tsx]: 👍 Removed usages of useRefAndState in favour of useState.

**Comment by AndrewPrifer** [comment]
I think one teeny-tiny improvement could be to leave some tolerance window for clicking, because right now if there's even just a little movement when clicking on the field, it won't become focused.

**Comment by AndrewPrifer** [comment]
Bug: with faster mouse movements, the value only increments tiny bits, and isn't reflective of the actual distance the pointer traveled.

**Review by AndrewPrifer — APPROVED** [review]
Looks good to me now!

**Pull request merged**